lib/outdated.js

```diff
@@ -1,437 +1,273 @@
 
-    return out.join(':')
-  }).join(os.EOL)
-}
+    if (opts.long) outHead.push('Package Type', 'Homepage')
+    const outTable = [outHead].concat(outList)
 
-function makeJSON (list, opts) {
-  var out = {}
-  list.forEach(function (p) {
-    var dep = p[0]
-    var depname = p[1]
-    var dir = dep.path
-    var has = p[2]
-    var want = p[3]
-    var latest = p[4]
-    var type = p[6]
-    if (!opts.global) {
-      dir = path.relative(process.cwd(), dir)
+    if (opts.color) {
+      outTable[0] = outTable[0].map(heading => styles.underline(heading))
     }
-    out[depname] = { current: has,
-      wanted: want,
-      latest: latest,
-      location: dir
-    }
-    if (long) {
-      out[depname].type = type
-      out[depname].homepage = dep.package.homepage
-    }
-  })
-  return JSON.stringify(out, null, 2)
-}
 
-function outdated_ (args, path, tree, parentHas, depth, opts, cb) {
-  if (!tree.package) tree.package = {}
-  if (path && moduleName(tree)) path += ' > ' + tree.package.name
-  if (!path && moduleName(tree)) path = tree.package.name
-  if (depth > opts.depth) {
-    return cb(null, [])
+    const tableOpts = {
+      align: ['l', 'r', 'r', 'r', 'l'],
+      stringLength: s => ansiTrim(s).length
+    }
+    output(table(outTable, tableOpts))
   }
-  var types = {}
-  var pkg = tree.package
 
-  if (!tree.children) tree.children = []
-
-  var deps = tree.error ? tree.children : tree.children.filter((child) => !isExtraneous(child))
+  process.exitCode = outdated.length ? 1 : 0
+}
 
-  deps.forEach(function (dep) {
-    types[moduleName(dep)] = 'dependencies'
-  })
+async function outdated_ (tree, deps, opts) {
+  const list = []
 
-  Object.keys(tree.missingDeps || {}).forEach(function (name) {
-    deps.push({
-      package: { name: name },
-      path: tree.path,
-      parent: tree,
-      isMissing: true
-    })
-    types[name] = 'dependencies'
-  })
+  const edges = new Set()
+  function getEdges(nodes, type) {
 
-  // If we explicitly asked for dev deps OR we didn't ask for production deps
-  // AND we asked to save dev-deps OR we didn't ask to save anything that's NOT
-  // dev deps then…
-  // (All the save checking here is because this gets called from npm-update currently
-  // and that requires this logic around dev deps.)
-  // FIXME: Refactor npm update to not be in terms of outdated.
-  var dev = opts.dev || /^dev(elopment)?$/.test(opts.also)
-  var prod = opts.production || /^prod(uction)?$/.test(opts.only)
-  if (
-    (dev || !prod) &&
-    (
-      opts['save-dev'] || (!opts.save && !opts['save-optional'])
-    )
-  ) {
-    Object.keys(tree.missingDevDeps).forEach(function (name) {
-      deps.push({
-        package: { name: name },
-        path: tree.path,
-        parent: tree,
-        isMissing: true
-      })
-      if (!types[name]) {
-        types[name] = 'devDependencies'
+    const getEdgesIn = (node) => {
+      for (const edge of node.edgesIn) {
+        edges.add({ name: edge.name, edge })
```

Comment:
It's not super relevant for performance or anything, since it's just a few thousand same-shaped POJO objects, but couldn't we just put the edges themselves in the set, since the `name` here will always be identical to `edge.name`?